src/agora/radio_calibrate.cc

```diff
@@ -602,55 +605,98 @@ bool RadioConfig::InitialCalib(bool sample_adjust) {
+        if (ret < (int)read_len) {
+          good_csi = false;
+          std::cout << "bad read (" << ret << ") at node " << ref
+                    << " from node " << i << std::endl;
+        }
+      }
+    }
+    // Transmit from Ref Antenna to Beamforming Antennas (Up)
+    {
       int tx_flags = SOAPY_SDR_WAIT_TRIGGER | SOAPY_SDR_END_BURST;
-      int ret = ba_stn_[i]->writeStream(this->tx_streams_[i], txbuff0.data(),
-                                        cfg_->PilotCi16().size(), tx_flags,
-                                        tx_time, 1000000);
+      int ret =
+          ba_stn_[ref]->writeStream(this->tx_streams_[ref], txbuff0.data(),
+                                    read_len, tx_flags, tx_time, 1000000);
       if (ret < (int)read_len) {
         std::cout << "bad write\n";
       }
-      for (size_t j = 0; j < r; j++) {
-        if (j == i) {
-          continue;
-        }
+
+      for (size_t i = 0; i < r; i++) {
+        if (i == ref) continue;
         int rx_flags = SOAPY_SDR_WAIT_TRIGGER | SOAPY_SDR_END_BURST;
-        ret = ba_stn_[j]->activateStream(this->rx_streams_[j], rx_flags,
+        ret = ba_stn_[i]->activateStream(this->rx_streams_[i], rx_flags,
                                          rx_time, read_len);
       }
 
       Go();
 
       int flags = 0;
-      for (size_t j = 0; j < r; j++) {
-        if (j == i) {
-          continue;
+      for (size_t i = 0; i < r; i++) {
+        if (good_csi == false) {
+          break;
         }
+        if (i == ref) continue;
         std::vector<void*> rxbuff(2);
-        rxbuff[0] = buff[(i * r + j)].data();
-        // rxbuff[1] = ant == 2 ? buff[(i*M+j)*ant+1].data() :
-        // dummyBuff.data();
-        // rxbuff[1 - ch] = dummybuff.data();
-        ret = ba_stn_[j]->readStream(this->rx_streams_[j], rxbuff.data(),
+        rxbuff[0] = buff[m + cfg_->NumChannels() * i].data();
+        if (cfg_->NumChannels() == 2)
+          rxbuff[1] = buff[m + cfg_->NumChannels() * i + 1].data();
+        ret = ba_stn_[i]->readStream(this->rx_streams_[i], rxbuff.data(),
                                      read_len, flags, rx_time, 1000000);
         if (ret < (int)read_len) {
           good_csi = false;
-          std::cout << "bad read (" << ret << ") at node " << j << " from node "
-                    << i << std::endl;
+          std::cout << "bad read (" << ret << ") at node " << i << " from node "
+                    << ref << std::endl;
         }
       }
     }
 
+    std::vector<std::vector<std::complex<int16_t>>> noise_buff;
+    noise_buff.resize(m);
     for (size_t i = 0; i < r; i++) {
       int rx_flags = SOAPY_SDR_END_BURST;
       int flags = 0;
       int ret = ba_stn_[i]->activateStream(this->rx_streams_[i], rx_flags,
                                            rx_time, read_len);
       std::vector<void*> rxbuff(2);
-      rxbuff[0] = buff[(i * r + i)].data();
+      noise_buff[cfg_->NumChannels() * i].resize(read_len);
```

Comment:
Since this is a vector, you might as well call the .at() function.  It could help in the future if we have overruns.